multi-module-app part 2: classpath resources

xabolcs opened the pull request:
This is a continuation of #291's second point:
> adds support for loading conf, routes **etc**. from JARs (technically, from classpath), not only from local folders.

The "etc" emphasize is mine! :slightly_smiling_face: 
In the added `hello.html` there was a TODO:
```html
  <title>Hello, multi-module RePlay app</title>
<!--  TODO Including files from classpath is not implemented yet -->
<!--  # {include 'icons/favicon.html'/}-->
<!--  # {include './meta.html'/}-->
```

So I started to implement those missing bits!

More TODOs:
* :heavy_check_mark: legacy (Play! 1.x) `FastTags` on classpath
* :heavy_check_mark: exercise `Router.reverseWithCheck()`'s [callers](https://github.com/search?q=repo%3Areplay-framework%2Freplay+__reverseWithCheck&type=code): <del>`ExecutableTemplate`</del> (#525), `GTJavaBase1xImpl` in `fastergt` module
* :fast_forward: exercise `precompiled/templates` on classpath
  * and make it constant :wink:  
  * this topic needs discussion as lot of precompiling stuff was removed (e.g. in commit 185ca449d2d1cf656f3aed973625b4a0f0bea3f8)
* ... ?

xabolcs commented:
Wow!
`netty3` succeeded, but `javanet` failed. :thinking: 
Sure I tested with `netty4` which is cancelled here.

Ahh, it's unable to save stream. :confused: 

Selenide:
```
download http://localhost:45997/ in 4000 ms.
Caused by: java.net.SocketTimeoutException: Read timed out
	at app//com.codeborne.selenide.impl.DownloadFileWithHttpRequest.download(DownloadFileWithHttpRequest.java:110)
	at app//com.codeborne.selenide.impl.DownloadFileWithHttpRequest.download(DownloadFileWithHttpRequest.java:91)
	at app//com.codeborne.selenide.SelenideDriver.download(SelenideDriver.java:394)
	at app//com.codeborne.selenide.SelenideDriver.download(SelenideDriver.java:382)
	at app//com.codeborne.selenide.Selenide.download(Selenide.java:848)
	at app//ui.hello.HelloWorldSpec.openHelloWorldPageAndCheckMeta(HelloWorldSpec.java:34)
	at java.base@17.0.13/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

Multi-module-app:
```
02:40:06,167 TRACE [play.mvc.ActionInvoker] ~ ------- public play.rebel.View controllers.HelloWorld.hello()
02:40:06,172 TRACE [play.server.javanet.PlayHandler$JavaNetInvocation] ~ init: end true
02:40:10,222 INFO  [com.codeborne.selenide.WebDriverThreadLocalContainer] ~ No webdriver is bound to current thread: 1 - let's create a new webdriver
02:40:10,623 TRACE [play.server.javanet.PlayHandler$JavaNetInvocation] ~ onSuccess: begin
02:40:10,623 TRACE [play.server.javanet.PlayHandler] ~ copyResponse: begin
02:40:10,624 TRACE [play.server.javanet.PlayHandler] ~ writeResponse: begin
02:40:10,624 TRACE [play.server.javanet.PlayHandler] ~ writeResponse: content length [279]
02:40:10,632 TRACE [play.server.javanet.PlayHandler] ~ serve500: begin
02:40:10,919 ERROR [play.server.javanet.PlayHandler] ~ Internal Server Error (500) for GET / (UnexpectedException)
play.exceptions.UnexpectedException: java.io.IOException: Broken pipe
	at play.Invocation.onActionInvocationException(Invocation.java:76)
	at play.server.javanet.PlayHandler$JavaNetInvocation.run(PlayHandler.java:416)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:62)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:132)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:97)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:53)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:532)
	at jdk.httpserver/sun.net.httpserver.Request$WriteStream.write(Request.java:424)
	at jdk.httpserver/sun.net.httpserver.FixedLengthOutputStream.write(FixedLengthOutputStream.java:81)
	at java.base/java.io.FilterOutputStream.write(FilterOutputStream.java:108)
	at jdk.httpserver/sun.net.httpserver.PlaceholderOutputStream.write(ExchangeImpl.java:458)
	at play.server.javanet.PlayHandler.writeResponse(PlayHandler.java:283)
	at play.server.javanet.PlayHandler.copyResponse(PlayHandler.java:151)
	at play.server.javanet.PlayHandler$JavaNetInvocation.onSuccess(PlayHandler.java:441)
	at play.server.javanet.PlayHandler$JavaNetInvocation.run(PlayHandler.java:413)
	... 6 more
	Suppressed: java.io.IOException: insufficient bytes written to stream
		at jdk.httpserver/sun.net.httpserver.FixedLengthOutputStream.close(FixedLengthOutputStream.java:92)
		at jdk.httpserver/sun.net.httpserver.PlaceholderOutputStream.close(ExchangeImpl.java:473)
		at play.server.javanet.PlayHandler.writeResponse(PlayHandler.java:282)
		... 9 more
02:40:10,924 ERROR [play.server.javanet.PlayHandler] ~ Error during the 500 response generation
java.io.IOException: headers already sent
	at jdk.httpserver/sun.net.httpserver.ExchangeImpl.sendResponseHeaders(ExchangeImpl.java:206)
	at jdk.httpserver/sun.net.httpserver.HttpExchangeImpl.sendResponseHeaders(HttpExchangeImpl.java:85)
	at play.server.javanet.PlayHandler.printResponse(PlayHandler.java:546)
	at play.server.javanet.PlayHandler.serve500(PlayHandler.java:564)
	at play.server.javanet.PlayHandler$JavaNetInvocation.run(PlayHandler.java:422)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
02:40:10,925 ERROR [play.server.javanet.PlayHandler] ~ serveStatic for request GET /
java.io.IOException: headers already sent
	at jdk.httpserver/sun.net.httpserver.ExchangeImpl.sendResponseHeaders(ExchangeImpl.java:206)
	at jdk.httpserver/sun.net.httpserver.HttpExchangeImpl.sendResponseHeaders(HttpExchangeImpl.java:85)
	at play.server.javanet.PlayHandler.printResponse(PlayHandler.java:546)
	at play.server.javanet.PlayHandler.sendServerError(PlayHandler.java:619)
	at play.server.javanet.PlayHandler.serve500(PlayHandler.java:579)
	at play.server.javanet.PlayHandler$JavaNetInvocation.run(PlayHandler.java:422)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
02:40:10,933 TRACE [play.server.javanet.PlayHandler] ~ serve500: end
02:40:10,934 TRACE [play.server.javanet.PlayHandler$JavaNetInvocation] ~ run: end

```

xabolcs commented on framework/src/log4j.properties:
```diff
@@ -1,5 +1,6 @@
 log4j.rootLogger=INFO, Console
 
+log4j.logger.play=TRACE
```
Ohh, it's slipped in! :grinning: 

Should remove before merge!

xabolcs commented:
The 4000ms timeout made it worse. :confused:

xabolcs commented on fastergt/src/play/modules/gtengineplugin/gt_integration/GTFileResolver1xImpl.java:
```diff
@@ -78,12 +80,32 @@ public GTTemplateLocationReal getTemplateLocationFromRelativePath(String relativ
 
     // TODO find in classpath?
     File vf = Play.file(relativePath);
```
Yep its a remnant, but I don't don't care for now!
Maybe I'll add a cleanup commit at the end. :slightly_smiling_face:

xabolcs commented on replay-tests/multi-module-app/app/build.gradle:
```diff
@@ -1,5 +1,6 @@
 dependencies {
   implementation project(':replay-tests:multi-module-app:core')
+  runtimeOnly project(':javanet')
```
I didn't test the other run configurations, so I don't know if they work at all or not.
Sure if those miss the http server backend, then yeah, putting into that `'../../replay-tests.gradle'` file would be better.

xabolcs replied:
Or you talking specifically about parameterizing the dependency (e.g. through system parameter)?

xabolcs replied:
> Yups. IIRC we already do that.

Just partially! It's a commented out `implementation` in `replay-tests.gradle`:
```gradle
dependencies {
  // To run the app locally:
  // implementation project(':netty3')

...
}
```

xabolcs commented on fastergt/src/play/templates/ExecutableTemplate.java:
```diff
@@ -139,7 +139,7 @@ public String __reverseWithCheck_absolute_false(String action) {
 
   private String __reverseWithCheck(String action, boolean absolute) {
-    return Router.reverseWithCheck(action, Play.file(action), absolute);
+    return Router.reverseWithCheck(action, play.server.ServerHelper.findFile(action), absolute);
```
@asolntsev , 

I think these `private private String __reverseWithCheck*` codes here in `play.templates.ExecutableTemplate` are dead codes! 
My reasoning is below.

In **Play! 1**:
```bash
$ git grep __reverseWithCheck
framework/src/play/templates/GroovyTemplate.java:        public String __reverseWithCheck_absolute_true(String action) {
framework/src/play/templates/GroovyTemplate.java:            return __reverseWithCheck(action, true);
framework/src/play/templates/GroovyTemplate.java:        public String __reverseWithCheck_absolute_false(String action) {
framework/src/play/templates/GroovyTemplate.java:            return __reverseWithCheck(action, false);
framework/src/play/templates/GroovyTemplate.java:        private String __reverseWithCheck(String action, boolean absolute) {
framework/src/play/templates/GroovyTemplateCompiler.java:                print("\tout.print(__reverseWithCheck_absolute_true(" + action + "));");
framework/src/play/templates/GroovyTemplateCompiler.java:                print("\tout.print(__reverseWithCheck_absolute_false(" + action + "));");
```
* it was defined in `play.templates.GroovyTemplate.ExecutableTemplate` class
* and was used by the compiler `play.templates.GroovyTemplateCompiler` class in method `protected void action(boolean absolute)`


In **Re-Play**:
```bash
$ git grep __reverseWithCheck
fastergt/src/play/modules/gtengineplugin/gt_integration/GTJavaBase1xImpl.java:  public String __reverseWithCheck_absolute_true(String action) {
fastergt/src/play/modules/gtengineplugin/gt_integration/GTJavaBase1xImpl.java:    return __reverseWithCheck(action, true);
fastergt/src/play/modules/gtengineplugin/gt_integration/GTJavaBase1xImpl.java:  public String __reverseWithCheck_absolute_false(String action) {
fastergt/src/play/modules/gtengineplugin/gt_integration/GTJavaBase1xImpl.java:    return __reverseWithCheck(action, false);
fastergt/src/play/modules/gtengineplugin/gt_integration/GTJavaBase1xImpl.java:  public static String __reverseWithCheck(String action, boolean absolute) {
fastergt/src/play/modules/gtengineplugin/gt_integration/GTPreCompiler1xImpl.java:      code = " out.append(__reverseWithCheck(\"" + action + "\", " + absolute + "));\n";
fastergt/src/play/templates/ExecutableTemplate.java:  public String __reverseWithCheck_absolute_true(String action) {
fastergt/src/play/templates/ExecutableTemplate.java:    return __reverseWithCheck(action, true);
fastergt/src/play/templates/ExecutableTemplate.java:  public String __reverseWithCheck_absolute_false(String action) {
fastergt/src/play/templates/ExecutableTemplate.java:    return __reverseWithCheck(action, false);
fastergt/src/play/templates/ExecutableTemplate.java:  private String __reverseWithCheck(String action, boolean absolute) {
```
* fastergt was imported in one of the first commits, 1f6edbcadd836eb19ccc807561097047bf7fc518
* `play.templates.ExecutableTemplate` class was introduced by extracting from the removed `play.templates.GroovyTemplate` class in commit 185ca449d2d1cf656f3aed973625b4a0f0bea3f8
* later it was moved from `:framework` to `:fastergt` in #65

As the only usage (`play.templates.GroovyTemplateCompiler.action()`) was gone when `play.templates.ExecutableTemplate` introduced, I think this is dead code.

What do you think? Do you remember anything relevant? :grinning: 
If it's really dead code, may I push a removal of it?

asolntsev replied:
@xabolcs Hi. I see that method `play.modules.gtengineplugin.gt_integration.GTJavaBase1xImpl#__reverseWithCheck` is still used. I created PR https://github.com/replay-framework/replay/pull/524 that shows the usage.

But yes, I guess most of code in `ExecutableTemplate` is not used anymore. It was used in "default" groovy templates in Play1 which were replaced by "faster groovy templates" in Replay.

xabolcs replied:
> Hi. I see that method `play.modules.gtengineplugin.gt_integration.GTJavaBase1xImpl#__reverseWithCheck` is still used.

Yup! But this is `play.templates.ExecutableTemplate`.

Will check that #524 PR soon! :slightly_smiling_face: 
I also touch those `@{'...'}` usages in this PR, but with classpath resources, so they won't clash! :+1: 

> But yes, I guess most of code in `ExecutableTemplate` is not used anymore.

I'll prepare a PR for these `play.templates.ExecutableTemplate#__reverseWitCheck*()` removals. I'll leave the other unused code removal to others! :grinning:

xabolcs commented:
@asolntsev ,

I already added a test with custom `play.tmpDir` but wanted a second batch of test with `play.tmpDir=none`:
```
22:00:28,855 DEBUG [play.Play] ~ No tmp folder will be used (play.tmp is set to none)
```

How could I do this without duplicating all the test? Or at least, make only the minimal duplications. :thinking:

xabolcs commented on fastergt/src/play/modules/gtengineplugin/gt_integration/GTFileResolver1xImpl.java:
```diff
@@ -87,7 +87,7 @@ public GTTemplateLocationReal getTemplateLocationFromRelativePath(String relativ
     if (vf == null) {
       try {
-        final String classloadedPrefix = "/" + Play.tmpDir.getName() + "/";
+        final String classloadedPrefix = "/" + Play.tmpDir.getName() + "/" + TemplateLoader.CLASSPATH_LOADED_TEMPLATE_TMP_PATH_PREFIX;
```
This dies with NPE when `play.tmpDir=none` is set in the `application.conf`, but a fix is ready, just a few tests need to be cooked.

xabolcs commented on framework/src/play/mvc/Router.java:
```diff
@@ -67,6 +68,10 @@ public static List<Route> routes() {
+    // This automatically added route is needed to serve the static files extracted from classpath.
+    // The "public" postfix (!) on the end is to prevent unwanted leakage of classpath resource files!
+    // TODO: automatically expose routes (e.g. with "staticDir:" parsing or with application.conf settings from dependencies instead of hardcoded "public"
+    addRoute("GET", "/public/", "staticDir:" + Play.tmpDir.getName() + "/" + SERVER_HELPER_FIND_FILE_TMP_PATH_PREFIX + "public");
```
This dies with NPE when `play.tmpDir=none` is set in `application.conf`.
Fix is ready, only tests need to be created!

asolntsev commented:
> How could I do this without duplicating all the test? Or at least, make only the minimal duplications. 🤔

Maybe it's not worth that? Maybe just don't test custom `play.tmpDir`... :)

xabolcs commented:
This PR is now ready to review.

The commit order is the following (separated by hand):
```
$ git log --oneline master..HEAD | tac
4a62064 replay-tests: multi-module-app: let it run

9f959dd replay-tests: enable failing example for static file on classpath
13daf01 serveStatic: improve ServerHelper.findFile() with classpath

b0bb1b6 replay-tests: enable failing example for relative template on classpath
5c0801d fastergt: gt_integration: resolve relative templates from classpath

3bb79fe replay-tests: add (now fixed) example for classpath with custom tmpDir
142dedb framework: move the classpath-loaded templates to their own dir

794696b replay-tests: add failing example with '@' for resources on classpath
2988fc4 framework: fix Router.reverseWithCheck() calls with classpath files

dcc72a9 replay-tests: add failing example for allClassesInProject() on classpath
077fb38 replay-tests: add failing tests for FastTags and GTFastTag on classpath
aa66650 framework: add jar classes from "play.classes.scanJars" property

3c5a6ae replay-tests: add failing tests for @include with /conf prefix
9c87a64 framework: use "filePrefix" for relative config files @include
```

cies reviewed (approved):
lgtm!

xabolcs commented:
* **cies** approved these changes 5 days ago
* **asolntsev** approved these changes yesterday
* **asolntsev** assigned **xabolcs** yesterday

Thank you all, but there is no **merge button** for me, so it is up to you @asolntsev , @cies when you merge this! :+1:

asolntsev commented:
@xabolcs @cies Maybe it's time to release Replay 2.7.0?

xabolcs commented:
@asolntsev, talking about [2.7.0](https://github.com/replay-framework/replay/milestone/20), If you are interested, I could give spin for that `Properties` PR - addressing your comments.
Of course It's OK to me to pass that to 2.7.1.

asolntsev commented:
@xabolcs Yes, why not. That PR was also useful. ;)
